src/ty.md

```diff
@@ -17,110 +179,378 @@ tcx: TyCtxt<'tcx>
 ```
 
-As you can see, the `TyCtxt` type takes a lifetime parameter.
-During Rust compilation, we allocate most of our memory in
-**arenas**, which are basically pools of memory that get freed all at
-once. When you see a reference with a lifetime like `'tcx`,
-you know that it refers to arena-allocated data (or data that lives as
-long as the arenas, anyhow).
+As you can see, the `TyCtxt` type takes a lifetime parameter. When you see a reference with a
+lifetime like `'tcx`, you know that it refers to arena-allocated data (or data that lives as long as
+the arenas, anyhow).
+
+## Allocating and working with types
+
+To allocate a new type, you can use the various `mk_` methods defined on the `tcx`. These have names
+that correspond mostly to the various kinds of type variants. For example:
+
+```rust,ignore
+let array_ty = tcx.mk_array(elem_ty, len * 2);
+```
+
+These methods all return a `Ty<'tcx>` – note that the lifetime you get back is the lifetime of the
+innermost arena that this `tcx` has access to. In fact, types are always canonicalized and interned
+(so we never allocate exactly the same type twice) and are always allocated in the outermost arena
+where they can be (so, if they do not contain any inference variables or other "temporary" types,
+they will be allocated in the global arena). However, the lifetime `'tcx` is always a safe
+approximation, so that is what you get back.
+
+> NB. Because types are interned, it is possible to compare them for equality efficiently using `==`
+> – however, this is almost never what you want to do unless you happen to be hashing and looking
+> for duplicates. This is because often in Rust there are multiple ways to represent the same type,
+> particularly once inference is involved. If you are going to be testing for type equality, you
+> probably need to start looking into the inference code to do it right.
 
-### Allocating and working with types
+You can also find various common types in the `tcx` itself by accessing `tcx.types.bool`,
+`tcx.types.char`, etc (see `CommonTypes` for more).
 
-Rust types are represented using the `Ty<'tcx>` defined in the `ty`
-module (not to be confused with the `Ty` struct from [the HIR]). This
-is in fact a simple type alias for a reference with `'tcx` lifetime:
+## Beyond types: other kinds of arena-allocated data structures
+
+In addition to types, there are a number of other arena-allocated data structures that you can
+allocate, and which are found in this module. Here are a few examples:
+
+- [`Substs`][subst], allocated with `mk_substs` – this will intern a slice of types, often used to
+  specify the values to be substituted for generics (e.g. `HashMap<i32, u32>` would be represented
+  as a slice `&'tcx [tcx.types.i32, tcx.types.u32]`).
+- `TraitRef`, typically passed by value – a **trait reference** consists of a reference to a trait
+  along with its various type parameters (including `Self`), like `i32: Display` (here, the def-id
```

Comment:
First usage of def id. Good to point out it’s defined and discussed in depth later

Reply:
Done